Lab 3

Bazalii opened the pull request.

Bazalii committed:
feat: assign all necessary classes

Bazalii committed:
feat: realise all BackupStructure classes

- Assign all fields in these classes
- Realise several methods

Bazalii committed:
feat: assign methods in FileSystem interfaces

Bazalii committed:
fix: delete useless class and interface

Bazalii committed:
feat: realise all methods for work with Windows FS

Bazalii committed:
feat: realise emulation of FS for tests

Bazalii committed:
feat: add new exception

Bazalii committed:
feat: add several new methods for work with FS

- assign corresponding methods in IFileSystem
- realise three new methods in WindowsFileSystem

Bazalii committed:
feat: add new method in WindowsFileSystem

Bazalii committed:
feat: realise storage algorithms

- make ISavingAlgorithm abstract class
- add realisation of two storage algorithms:
  SplitStorage and SingleStorage

Bazalii committed:
feat: add new method to FS interface

Bazalii committed:
feat: some improvements in BackupJob

- add two new fields
- correct Process() method

Bazalii committed:
feat: add all necessary realisations for FS emulation

- add several interfaces with corresponding
  realisations for emulation of FS

Bazalii committed:
feat: improve GetPath() method in MemoryDirectory

Bazalii committed:
feat: change MemoryFile constructor to common format

Bazalii committed:
feat: improve SavingAlgorithm implementations

- Add new method SetBackupsDirectoryPath
- Change constructors of implementations to have
  an opportunity to set path there

Bazalii committed:
fix: rename exception

Bazalii committed:
fix: remove repeating code

Bazalii committed:
feat & fix: some improvements in MemoryFileSystem

- Add realisation of ExtractFromArchive method
- Add realisation of CreateArchive method
- Change DepthFirstSearch method to make it work properly
- Correct CreateFile, CopyFile and CreateDirectory methods

Bazalii committed:
feat: add Equals methods to JobObject class

- Now JobObject can be conveniently removed from List<JobObject>

Bazalii committed:
feat: improve BackupJob

- Add three new methods
- Add new field for saving path to Backups

Bazalii committed:
feat: improve classes of FS emulation

- Add Equals method for convenient deletion
- Change constructor in MemoryArchive
- Add RemoveObject method in MemoryArchive

Bazalii committed:
feat: add tests for Backups

Bazalii committed:
fix: remove save calls

- There is no need in them because if the object
  hasn't been found exception will be thrown

Bazalii committed:
fix: remove unused usages

Bazalii committed:
fix: minor changes in RestorePoint

- Remove setter from TimeOfBackup
- Make _backupedFiles readonly
- Change argument of constructor to IEnumerable

Bazalii committed:
feat: add new exception

- This exception will be used in MemoryFileSystem
  to check if StorageObject is already made

Bazalii committed:
feat & fix: changes in FS classes

- Add usage of range indexer
- Rename all abstract classes: ISmth -> Smth and
  Directory -> VirtualDirectory, File -> VirtualFile
- Add try catch constructions in CreateDirectory,
  CreateFile and CreateArchive methods to check
  if StorageObject already exists
- Add creation of an archive to AddToArchive method
  if the archive doesn't exist
- SavingAlgorithms now create directory when SetBackupsDirectoryPath
  method is called
- Fix bug in RemoveDirectory method in MemoryFileSystem class

Bazalii committed:
feat: correct tests so that they reflect FS changes

Bazalii committed:
fix: add null checks in constructors

hrrrrustic committed:
Merge branch 'master' into lab-3

Bazalii committed:
feat: add IArchiver interface and one implementation of it

- Realise WinZipArchiver: archiver for Win FS, which uses ZipFile

Bazalii committed:
feat: make MemoryFileSystem singleton

Bazalii committed:
feat: add property Archiver to SavingAlgorithm

- Add property of type IArchiver for creating archives
- Reflect changes in inheritors

Bazalii committed:
fix: remove archive methods from IFileSystem

Bazalii committed:
feat: add VirtualArchiver for MemoryFileSystem

Bazalii committed:
feat: improve tests

- Add new field for archiver
- Add usage of singleton and archiver

Bazalii committed:
fix: replace try catch constructions with if statements

Bazalii committed:
fix: make fields in BackupJob protected for future extension

ronimizy reviewed (changes requested):
Нужно изменить абстракцию архиватора, чтобы она использовала абстракцию над файловой системой при получении данных.
Быстрый фикс, потом апрув.

Bazalii committed:
feat: change logic of archiver

- Add usage of GZip library
- Now archiver just compresses or decompresses files and doesn't work with file system

Bazalii committed:
feat: improve FileSystems

- Realise method ReadFile for reading the content in files
- Change logic of WriteToFile method

Bazalii committed:
feat: change usage of archiver in algorithms

Bazalii committed:
fix: minor changes

- Remove unused System.Text
- Remove inheritance from IArchiver from VirtualArchiver
  because this class is just a kind of additional
  functionality for MemoryFileSystem

Bazalii committed:
fix: change tests according to the changes

ronimizy reviewed (approved):
![telegram-cloud-photo-size-2-5283028323947098510-c](https://user-images.githubusercontent.com/71693183/147139579-9e643140-6c02-45fd-9144-74ba9ac12a88.jpg)

Bazalii committed:
fix: delete unused functions
